### PR TITLE
Add uri protocol handler

### PR DESCRIPTION
**Pull request by IntelOrca**
Implementing #2691 for Windows.

Currently supports joining a server:
`openrct2://join/127.0.0.1:11753`

### CODE REVIEW

**Comment by IntelOrca** [comment]
@janisozaur @marijnvdwerf @LRFLEW 

Well this is it done for Windows, I have no idea how it is done for other platforms.

**Comment by LRFLEW** [comment]
Hmm.. so on Windows it passes it as a command-line argument. How does it handle it if the application is already open? Does it create a new instance of the game?

On macOS, it really doesn't like having multiple instances of an application open, so Apple used an event-based system, where it appears that you register a listener for the scheme (and declare it in the Info.plist file). This should be simple enough to implement. On iOS, the even is triggered in the Application Delegate, which is controlled by SDL, so if we ever want to implement it here, we will have some issues.

**Comment by IntelOrca** [comment]
@LRFLEW At the moment it just starts a new instance.

**Comment by marijnvdwerf** [comment]
What @LRFLEW means to say is that it only starts a new instance when the app is not running. If the app was already running, it doesn't attempt to start a new instance.

**Comment by LRFLEW** [comment]
I think I have it working. Since it uses an event on macOS, I ended up using SDL2's user event system to get the state to change on the main thread. I'll PR @IntelOrca's branch in a few minutes.

**Comment by zsilencer** [comment]
would be good to get this to work with ipv6 address:port

**Comment by LRFLEW** [comment]
Is this not? AFAIK, because ipv6 addresses uses the colon to separate the parts of the address, you have to put the ipv6 address in brackets (ex, `[fe80::1]:12345`). I know there's code somewhere in the codebase that checks for and handles the brackets.

**Comment by LRFLEW** [comment]
Side note: I think it would be a good idea to add support for a `+` in the IP list to allow for multiple options. For example: `openrct2://join/192.168.0.2:12345+[fe80::2]:12345`. I don't know if we have a system in place to support multiple addresses for servers (or if it supports DNS names for servers), but this could be useful if we want to support that.

**Comment by IntelOrca** [comment]
The master server supports multiple addresses, but the game doesn't I didn't even think the game supported IPv6. Or at least I don't think the master server does - despite having it in the JSON.

File: src/openrct2/platform/windows.c [hunk 1]
```diff
@@ -1213,7 +1213,12 @@ bool platform_setup_uri_protocol()
 						HKEY hMuiCacheKey;
 						if (RegCreateKeyW(hRootKey, MUI_CACHE, &hMuiCacheKey) == ERROR_SUCCESS) {
 							swprintf_s(buffer, sizeof(buffer), L"%s.FriendlyAppName", exePath);
+#ifdef __MINGW32__
+							// MINGW defines RegSetKeyValueW's signature incorrectly
+							RegSetKeyValueW(hMuiCacheKey, NULL, (LPCSTR)buffer, REG_SZ, (LPCSTR)L"OpenRCT2", sizeof(L"OpenRCT2") + 1);
```
Comment (janisozaur) [inline src/openrct2/platform/windows.c]: Can you please leave a note that this refers to mingw-w64 (apparently there is more than one) and that a fix was submitted upstream, so it should be removed when we can use their next release: https://sourceforge.net/p/mingw-w64/mingw-w64/ci/da9341980a4b70be3563ac09b5927539e7da21f7/

File: src/openrct2/platform/windows.c [hunk 2]
```diff
@@ -1040,7 +1040,8 @@ utf8* platform_get_username() {
 // File association setup
 ///////////////////////////////////////////////////////////////////////////////
 
-#define SOFTWARE_CLASSES L"Software\\Classes"
+#define SOFTWARE_CLASSES	L"Software\\Classes"
+#define MUI_CACHE			L"Local Settings\\Software\\Microsoft\\Windows\\Shell\\MuiCache"
```
Comment (janisozaur) [inline src/openrct2/platform/windows.c]: misaligned

Comment (IntelOrca) [inline src/openrct2/platform/windows.c]: they are not misaligned

**Pull request merged**